### PR TITLE
OVS HWOL: don't rename VF rep to container ID

### LINKED ISSUE #1933: ovs hw offload: rename vf rep to its original one
VF repo is renamed with last 15 bits of container ID when attaching to ovs bridge, but it is not restored when released from pod/container, this will result in the vf representor name being a container ID that gets removed.
Possible fix might be: Cache the original vf rep name and restore upon pod deletion (cmdDel)

### PR DESCRIPTION
**Pull request by zshi-redhat**
Fixes #1933

Signed-off-by: Zenghui Shi <zshi@redhat.com>

### CODE REVIEW

**Review by moshe010 — COMMENTED** [review]
The change look good to me. How well did you test it? I remember someone said that the code rely on the interface name to be the containerID, but if that not an issue you have my LGTM

**Comment by zshi-redhat** [comment]
> The change look good to me. How well did you test it? I remember someone said that the code rely on the interface name to be the containerID, but if that not an issue you have my LGTM

I tested #2005 where VF repo is not renamed, it works. That inspired me to change the regular ovs hw offload.
I will test this with CX-5 and update the result.

**Comment by adrianchiris** [comment]
> The change look good to me. How well did you test it? I remember someone said that the code rely on the interface name to be the containerID, but if that not an issue you have my LGTM

It relies on it for `PlatformSpecificCleanup()` on cmdDel. in smartNIC case the cleanup is done on SmartNIC and not by cni server.
the representor in smartNIC case is derrived from pf/vf id and not sandbox ID.

so i think in hw offload non-smart NIC case the representor will remain plugged to the bridge.

**Comment by adrianchiris** [comment]
to avoid rename: on cleanup need to derive rep name from vf netdev and use it to delete the port, or alternatively rely on the ovs interface `external-ids` column and delete the port whose sandbox field equals the Pod's `SandboxID`

heres an example of an `ovs-vsctl get interface <vf-rep> external-ids`

```
{attached_mac="0a:58:c0:a8:01:06", iface-id=default_bf-curl-pod-vd4vj, ip_addresses="192.168.1.6/24", sandbox=d73395441f13ceef2e87195c48bf0710efecc391bc32ca076c7e8260396e1f45}
```

**Comment by zshi-redhat** [comment]
> to avoid vf rep rename you could:
> 
> On cleanup, derive rep name from vf netdev and use it to delete the port.
> Alternatively rely on the ovs interface `external-ids` column and delete the port whose `sandbox` field equals the Pod's `SandboxID`
> 
Thanks @adrianchiris ! 
Updated to use `external-ids:sandbox` to find vf rep as ifName.

File: go-controller/pkg/cni/helper_linux.go
Comment (adrianchiris) [inline go-controller/pkg/cni/helper_linux.go:414]: there is a slight chance that somehow the VF representor will be removed/renamed, in that case ovsFind will fail to find the Interface. 
in this case, i think you should catch the error and log it.

In addition, if we do hit this case we will have an OVS port without an interface right ? are you relying on the `checkForStaleOVSInterfaces()` to remove the port from the bridge ?

**Comment by girishmg** [comment]
@zshi-redhat @adrianchiris 

1. When we add the port corresponding to the Pod  to `br-int` we don't use `--if-exists`,  so if the VF representor wasn't removed from `br-int` successfully on cmdDel(), then the cmdAdd() will fail next time around.

```
	ovsArgs := []string{
		"add-port", "br-int", hostIface.Name, "--", "set",
		"interface", hostIface.Name,
		fmt.Sprintf("external_ids:attached_mac=%s", ifInfo.MAC),
		fmt.Sprintf("external_ids:iface-id=%s", ifaceID),
		fmt.Sprintf("external_ids:ip_addresses=%s", strings.Join(ipStrs, ",")),
		fmt.Sprintf("external_ids:sandbox=%s", pr.SandboxID),
	}
```

2. BTW, this change is not backwards compatible. If someone does rolling update to your changes, then all of the existing pod's VF representor will not be successfully removed from `br-int` and they won't be renamed back to the VF representor name for use in subsequent cmdAdd()

3. cniCheck() will also fail on rolling update
```
# cat ovn-kubernetes-d83b6208d8d62cbf88e7bbd4f654cadab3fe2d803d6a659f9b684350c3957a8b-eth0 |jq .
{
  "cniVersion": "0.4.0",
  "interfaces": [
    {
      "name": "d83b6208d8d62cb",  <------ this is on the flie system.
      "mac": "02:3e:83:10:fc:8e"
    },
    {
      "name": "eth0",
      "mac": "0a:58:0a:c0:25:10",
      "sandbox": "/proc/2852612/ns/net"
    }
  ],
```

Not sure, how critical it is to use representor name? We can keep it consistent with the non-VF case like we have today, but still fix the issue of renaming the representor name from `sandobxID{:15]` to `VF_representor name.`

File: go-controller/pkg/cni/helper_linux.go
Comment (girishmg) [reply go-controller/pkg/cni/helper_linux.go:414]: @adrianchris unfortunately we can't depend on checkForStaleOVSInterfaces() since that funciton only checks for `ofport` that is set to `-1`. 

If cmdDel() fails to remove the port from `br-int`, then the `ofport` will not be `-1`. It will be a non-negative number and will not be cleared by checkForStaleOVSInterfaces()

Comment (adrianchiris) [reply go-controller/pkg/cni/helper_linux.go:414]: I think you refer to a different (perfectly valid) case where the interface exists in the interface table but the delete port command below fails.

I was referring to a scenario where the ovsFind fails to find the interface, meaning the ofport is set to -1.

1. connect vf rep to bridge
```
# ovs-vsctl add-port br-int pf1vf5
# ovs-vsctl -- --columns=name,ofport list Interface 
....
....                                                                                                                                                                                                                                                                                                                                                          
name                : pf1vf5                                                                       ​
ofport              : 7 
```

2. rename vf rep
```
# ip link set pf1vf5 down
# ip link set pf1vf5 name pf1vf55
# ovs-vsctl -- --columns=name,ofport list Interface 
....
....                                                                                                                                                                                                                                                                                                                                                                                               
name                : pf1vf5                                                                                                                                                                  
ofport              : -1 
```

**Comment by adrianchiris** [comment]
> BTW, this change is not backwards compatible. If someone does rolling update to your changes, then all of the existing pod's VF representor will not be successfully removed from br-int and they won't be renamed back to the VF representor name for use in subsequent cmdAdd()

You mean if someone does a rolling downgrade ? old ovnkube node will not be able to remove the non-renamed vf rep ports ?
because new logic will work on vf rep ports that have been renamed by old ovnkube node as it relies on the name as stored in ovs Interface table.

**Comment by girishmg** [comment]
> > BTW, this change is not backwards compatible. If someone does rolling update to your changes, then all of the existing pod's VF representor will not be successfully removed from br-int and they won't be renamed back to the VF representor name for use in subsequent cmdAdd()
> 
> You mean if someone does a rolling downgrade ? old ovnkube node will not be able to remove the non-renamed vf rep ports ?
> because new logic will work on vf rep ports that have been renamed by old ovnkube node as it relies on the name as stored in ovs Interface table.

Correct. IMHO, what we currently have is fine since it would work the same for VF and VETH case.

**Comment by girishmg** [comment]
> > Not sure, how critical it is to use representor name?
> 
> Not critical, this is for easy debugging. For example, when there are tens of pods' interfaces attached to br-int, it is not easy to tell which interface is VF, which is veth. sometimes, user may want to dump traffic from VF to confirm if packet is offloaded or not.
> 
> > We can keep it consistent with the non-VF case like we have today, but still fix the issue of renaming the representor name from `sandobxID{:15]` to `VF_representor name.`
> 
> @girishmg do you mean rename VF rep back to its original name when released from Pod?

Yeah, the initial issue reported in PR#1933. Thanks.

File: go-controller/pkg/cni/types.go
Comment (adrianchiris) [inline go-controller/pkg/cni/types.go:21]: nit: empty commment

File: go-controller/pkg/cni/helper_linux.go
Comment (adrianchiris) [inline go-controller/pkg/cni/helper_linux.go:107]: should this be saved in json format so if later other information need to cached then extending this would be easier ? (e.g to restore VF name as well)

Comment (zshi-redhat) [reply go-controller/pkg/cni/helper_linux.go:107]: I can add the -default suffix, how can we tell it's a default or additional when multiple networks are supported?

Comment (zshi-redhat) [reply go-controller/pkg/cni/helper_linux.go:108]: Done

File: go-controller/pkg/cni/types.go
Comment (zshi-redhat) [reply go-controller/pkg/cni/types.go]: Done

**Review by trozet — CHANGES_REQUESTED** [review]
As we talked about on slack, rather than using a file I think a better solution is to annotate the pods with the old VF name. Then we can keep a map[pods]previousVFName in CNI server. On ovnkube-node reboot, during CNIServer Init, we can read all the pods and rebuild the map cache from the annotations. Maybe @dcbw should weigh in too with his opinion on this.

**Comment by zshi-redhat** [comment]
> As we talked about on slack, rather than using a file I think a better solution is to annotate the pods with the old VF name. Then we can keep a map[pods]previousVFName in CNI server. On ovnkube-node reboot, during CNIServer Init, we can read all the pods and rebuild the map cache from the annotations. Maybe @dcbw should weigh in too with his opinion on this.

On pod cleanup, do we prefer to do the renaming in `cmdDel` or have a separate watch controller like how smartnic is handled: https://github.com/ovn-org/ovn-kubernetes/pull/2005/files#diff-d91e350d13922202c7af954976fe514b2ac18de2ba91bbf5d46a6b7e95fcc8f6R20

@adrianchiris by adding non-smartnic sriov info in pod annotation, I expect there will be some similarity on how the annotation structure is defined, do you think we should converge the smartnic and general sriov case into one single annotation structure?

For example, we may just need to add a vfRepName here for general sriov case: https://github.com/ovn-org/ovn-kubernetes/pull/2005/files#diff-57fb62e1da38d64d7ce05272e257730ab43969362c7236b697b372179dd47cdfR46-R49

**Comment by zshi-redhat** [comment]
> Hi @zshi-redhat, what is the status of this MR?

@DmytroLinkin  Sorry for the late response. I'm closing this PR as I don't find time working on this. Feel free to create another PR for it if you think this is needed.
cc @wizhaoredhat @bn222 FYI.

**Pull request closed without merging**
